File: python/de_message_parser_base.py
"""
Python implementation of de_common/de_databus/de_message_parser_base.hpp/cpp
Equivalent to C++ CAndruavMessageParserBase class
"""

import json
from abc import ABC, abstractmethod
from typing import Dict, Any, Union
try:
    from .configFile import ConfigFile
    from .localConfigFile import LocalConfigFile
    from .messages import *
except ImportError:
    from configFile import ConfigFile
    from localConfigFile import LocalConfigFile
    from messages import *


class AndruavMessageParserBase(ABC):
    """
    Python equivalent of C++ CAndruavMessageParserBase class.
    Abstract base class for parsing Andruav messages.
    """
    
    def __init__(self):
        self._is_binary = False
        self._is_system = False
        self._is_inter_module = False
        self._facade = None  # Will be set by subclass
    
    def parse_message(self, andruav_message: Dict[str, Any], full_message: Union[str, bytes], full_message_length: int):
        """
        Parse incoming Andruav message
        """
        message_type = andruav_message.get(ANDRUAV_PROTOCOL_MESSAGE_TYPE, 0)
        
        # Check if message is binary (doesn't end with '}')
        if isinstance(full_message, bytes):
            self._is_binary = True
        else:
            self._is_binary = not (full_message.endswith('}') or 
                                  (len(full_message) > 1 and full_message[-2] == '}'))
        
        # Get permission if available
        permission = 0
        if self._validate_field(andruav_message, ANDRUAV_PROTOCOL_MESSAGE_PERMISSION, int):
            permission = int(andruav_message[ANDRUAV_PROTOCOL_MESSAGE_PERMISSION])
        
        # Check if message is from system
        self._is_system = False
        if (self._validate_field(andruav_message, ANDRUAV_PROTOCOL_SENDER, str) and
            andruav_message[ANDRUAV_PROTOCOL_SENDER] == ANDRUAV_PROTOCOL_SENDER_COMM_SERVER):
            self._is_system = True
        
        # Check if message is inter-module
        self._is_inter_module = False
        if (self._validate_field(andruav_message, INTERMODULE_ROUTING_TYPE, str) and
            andruav_message[INTERMODULE_ROUTING_TYPE] == CMD_TYPE_INTERMODULE):
            self._is_inter_module = True
        
        # Handle RemoteExecute messages
        if message_type == TYPE_AndruavMessage_RemoteExecute:
            self.parse_remote_execute(andruav_message)
            return
        
        # Parse default commands and custom commands
        self._parse_default_command(andruav_message, full_message, full_message_length, message_type, permission)
        self.parse_command(andruav_message, full_message, full_message_length, message_type, permission)
    
    def _parse_default_command(self, andruav_message: Dict[str, Any], full_message: Union[str, bytes], 
                              full_message_length: int, message_type: int, permission: int):
        """
        Parse default commands that are common across all modules
        """
        cmd = andruav_message.get(ANDRUAV_PROTOCOL_MESSAGE_CMD, {})
        
        if message_type == TYPE_AndruavMessage_CONFIG_ACTION:
            self._handle_config_action(andruav_message, cmd)
    
    def _handle_config_action(self, andruav_message: Dict[str, Any], cmd: Dict[str, Any]):
        """
        Handle configuration action commands
        """
        module_key = ""
        
        # Validate action field
        if not self._validate_field(cmd, "a", int):
            return
        
        # Check module key if specified
        if self._validate_field(cmd, "b", str):
            module_key = self._get_module_key()
            if module_key != cmd["b"]:
                return
        
        action = cmd["a"]
        
        if action == CONFIG_ACTION_Restart:
            exit(0)
        
        elif action == CONFIG_ACTION_APPLY_CONFIG:
            config = cmd.get("c", {})
            print(config)
            config_file = ConfigFile.get_instance()
            config_file.update_json(json.dumps(config))
        
        elif action == CONFIG_REQUEST_FETCH_CONFIG_TEMPLATE:
            if ANDRUAV_PROTOCOL_SENDER not in andruav_message:
                return
            
            sender = andruav_message[ANDRUAV_PROTOCOL_SENDER]
            
            print(f"\n\033[96mCONFIG_REQUEST_FETCH_CONFIG_TEMPLATE\033[0m")
            
            try:
                with open("template.json", 'r') as file:
                    file_content = file.read()
                    file_content_json = json.loads(file_content)
                    self._facade.api_send_config_template(sender, module_key, file_content_json, True)
            except (IOError, json.JSONDecodeError):
                print(f"\n\033[1;91mcannot read template.json\033[0m")
                if self._facade:
                    self._facade.send_error_message("", 0, ERROR_3DR, 
                                                 NOTIFICATION_TYPE_ERROR, "cannot read template.json")
                    empty_file_content_json = {}
                    self._facade.api_send_config_template(sender, module_key, empty_file_content_json, True)
                return
        
        elif action == CONFIG_REQUEST_FETCH_CONFIG:
            # Implementation for fetching current config
            pass
        
        # Add other config actions as needed
    
    def _validate_field(self, json_obj: Dict[str, Any], field_name: str, expected_type: type) -> bool:
        """
        Validate if a field exists and has the expected type
        """
        return (field_name in json_obj and 
                isinstance(json_obj[field_name], expected_type))
    
    def _get_module_key(self) -> str:
        """
        Get module key - should be implemented by subclass or injected dependency
        """
        # This would typically come from a module instance
        # For now, return empty string
        return ""
    
    @property
    def is_binary(self) -> bool:
        """Check if message is binary"""
        return self._is_binary
    
    @property
    def is_system(self) -> bool:
        """Check if message is from system"""
        return self._is_system
    
    @property
    def is_inter_module(self) -> bool:
        """Check if message is inter-module"""
        return self._is_inter_module
    
    @abstractmethod
    def parse_remote_execute(self, andruav_message: Dict[str, Any]):
        """
        Parse remote execute commands - must be implemented by subclass
        """
        pass
    
    @abstractmethod
    def parse_command(self, andruav_message: Dict[str, Any], full_message: Union[str, bytes], 
                    full_message_length: int, message_type: int, permission: int):
        """
        Parse custom commands - must be implemented by subclass
        """
        pass


# Export the class for compatibility with C++ naming convention
CAndruavMessageParserBase = AndruavMessageParserBase

File: nodejs/de_message_parser_base.js
/**
 * NodeJS implementation of de_common/de_databus/de_message_parser_base.hpp/cpp
 * Equivalent to C++ CAndruavMessageParserBase class
 */

const fs = require('fs');
const {
    ANDRUAV_PROTOCOL_MESSAGE_TYPE,
    ANDRUAV_PROTOCOL_MESSAGE_PERMISSION,
    ANDRUAV_PROTOCOL_SENDER,
    INTERMODULE_ROUTING_TYPE,
    CMD_TYPE_INTERMODULE,
    ANDRUAV_PROTOCOL_SENDER_COMM_SERVER,
    TYPE_AndruavMessage_RemoteExecute,
    TYPE_AndruavMessage_CONFIG_ACTION,
    CONFIG_ACTION_Restart,
    CONFIG_ACTION_APPLY_CONFIG,
    CONFIG_REQUEST_FETCH_CONFIG_TEMPLATE,
    CONFIG_REQUEST_FETCH_CONFIG,
    ERROR_3DR,
    NOTIFICATION_TYPE_ERROR,
    ANDRUAV_PROTOCOL_MESSAGE_CMD,
    TYPE_AndruavMessage_Error,
    TYPE_AndruavMessage_CONFIG_STATUS,
    CONFIG_STATUS_FETCH_CONFIG_TEMPLATE
} = require('./messages');

class AndruavMessageParserBase {
    /**
     * Abstract base class for parsing Andruav messages
     */
    constructor() {
        this._is_binary = false;
        this._is_system = false;
        this._is_inter_module = false;
        this._facade = null; // Will be set by subclass
    }

    /**
     * Parse incoming Andruav message
     * @param {Object} andruav_message - Parsed JSON message
     * @param {string|Buffer} full_message - Raw message
     * @param {number} full_message_length - Message length
     */
    parseMessage(andruav_message, full_message, full_message_length) {
        const messageType = andruav_message[ANDRUAV_PROTOCOL_MESSAGE_TYPE] || 0;
        
        // Check if message is binary (doesn't end with '}')
        if (Buffer.isBuffer(full_message)) {
            this._is_binary = true;
        } else {
            const str = full_message.toString();
            this._is_binary = !(str.endsWith('}') || (str.length > 1 && str[str.length - 2] === '}'));
        }

        // Get permission if available
        let permission = 0;
        if (this._validateField(andruav_message, ANDRUAV_PROTOCOL_MESSAGE_PERMISSION, 'number')) {
            permission = andruav_message[ANDRUAV_PROTOCOL_MESSAGE_PERMISSION];
        }

        // Check if message is from system
        this._is_system = false;
        if (this._validateField(andruav_message, ANDRUAV_PROTOCOL_SENDER, 'string') &&
            andruav_message[ANDRUAV_PROTOCOL_SENDER] === ANDRUAV_PROTOCOL_SENDER_COMM_SERVER) {
            this._is_system = true;
        }

        // Check if message is inter-module
        this._is_inter_module = false;
        if (this._validateField(andruav_message, INTERMODULE_ROUTING_TYPE, 'string') &&
            andruav_message[INTERMODULE_ROUTING_TYPE] === CMD_TYPE_INTERMODULE) {
            this._is_inter_module = true;
        }

        // Handle RemoteExecute messages
        if (messageType === TYPE_AndruavMessage_RemoteExecute) {
            this.parseRemoteExecute(andruav_message);
            return;
        }

        // Parse default commands and custom commands
        this._parseDefaultCommand(andruav_message, full_message, full_message_length, messageType, permission);
        this.parseCommand(andruav_message, full_message, full_message_length, messageType, permission);
    }

    /**
     * Parse default commands that are common across all modules
     * @private
     */
    _parseDefaultCommand(andruav_message, full_message, full_message_length, messageType, permission) {
        const cmd = andruav_message[ANDRUAV_PROTOCOL_MESSAGE_CMD] || {};

        if (messageType === TYPE_AndruavMessage_CONFIG_ACTION) {
            this._handleConfigAction(andruav_message, cmd);
        }
    }

    /**
     * Handle configuration action commands
     * @private
     */
    _handleConfigAction(andruav_message, cmd) {
        let module_key = "";
        
        // Validate action field
        if (!this._validateField(cmd, "a", 'number')) {
            return;
        }

        // Check module key if specified
        if (this._validateField(cmd, "b", 'string')) {
            module_key = this._getModuleKey();
            if (module_key !== cmd["b"]) {
                return;
            }
        }

        const action = cmd["a"];

        if (action === CONFIG_ACTION_Restart) {
            console.log("CONFIG_ACTION_Restart - exiting");
            process.exit(0);
        }

        else if (action === CONFIG_ACTION_APPLY_CONFIG) {
            const config = cmd["c"] || {};
            console.log("CONFIG_ACTION_APPLY_CONFIG:", config);
            // Note: ConfigFile implementation would go here
            // For now, just print the config
        }

        else if (action === CONFIG_REQUEST_FETCH_CONFIG_TEMPLATE) {
            if (!andruav_message.hasOwnProperty(ANDRUAV_PROTOCOL_SENDER)) {
                return;
            }

            const sender = andruav_message[ANDRUAV_PROTOCOL_SENDER];
            console.log("\n\x1b[96mCONFIG_REQUEST_FETCH_CONFIG_TEMPLATE\x1b[0m");

            try {
                const fileContent = fs.readFileSync("template.json", 'utf8');
                const fileContentJson = JSON.parse(fileContent);
                if (this._facade) {
                    this._facade.api_send_config_template(sender, module_key, fileContentJson, true);
                }
            } catch (error) {
                console.log("\n\x1b[1;91mcannot read template.json\x1b[0m");
                if (this._facade) {
                    this._facade.send_error_message("", 0, ERROR_3DR, 
                        NOTIFICATION_TYPE_ERROR, "cannot read template.json");
                    const emptyFileContentJson = {};
                    this._facade.api_send_config_template(sender, module_key, emptyFileContentJson, true);
                }
                return;
            }
        }

        else if (action === CONFIG_REQUEST_FETCH_CONFIG) {
            // Implementation for fetching current config
            console.log("CONFIG_REQUEST_FETCH_CONFIG - not implemented");
        }

        // Add other config actions as needed
    }

    /**
     * Validate if a field exists and has the expected type
     * @private
     */
    _validateField(jsonObj, fieldName, expectedType) {
        return jsonObj.hasOwnProperty(fieldName) && typeof jsonObj[fieldName] === expectedType;
    }

    /**
     * Get module key - should be implemented by subclass or injected dependency
     * @private
     */
    _getModuleKey() {
        // This would typically come from a module instance
        // For now, return empty string
        return "";
    }

    /**
     * Check if message is binary
     * @returns {boolean}
     */
    get isBinary() {
        return this._is_binary;
    }

    /**
     * Check if message is from system
     * @returns {boolean}
     */
    get isSystem() {
        return this._is_system;
    }

    /**
     * Check if message is inter-module
     * @returns {boolean}
     */
    get isInterModule() {
        return this._is_inter_module;
    }

    /**
     * Parse remote execute commands - must be implemented by subclass
     * @abstract
     */
    parseRemoteExecute(andruav_message) {
        throw new Error("parseRemoteExecute must be implemented by subclass");
    }

    /**
     * Parse custom commands - must be implemented by subclass
     * @abstract
     */
    parseCommand(andruav_message, full_message, full_message_length, messageType, permission) {
        throw new Error("parseCommand must be implemented by subclass");
    }
}

// Export the class for compatibility with C++ naming convention
module.exports = {
    AndruavMessageParserBase,
    CAndruavMessageParserBase: AndruavMessageParserBase  // Alias for C++ compatibility
};
